packages/cloud_firestore/cloud_firestore_platform_interface/lib/src/types/query.dart

```diff
@@ -0,0 +1,27 @@
+/// An enumeration of firestore source types.
+enum Source {
+  /// Causes Firestore to try to retrieve an up-to-date (server-retrieved) snapshot, but fall back to
+  /// returning cached data if the server can't be reached.
+  serverAndCache,
+
+  /// Causes Firestore to avoid the cache, generating an error if the server cannot be reached. Note
+  /// that the cache will still be updated if the server request succeeds. Also note that
+  /// latency-compensation still takes effect, so any pending write operations will be visible in the
+  /// returned data (merged into the server-provided data).
+  server,
+
+  /// Causes Firestore to immediately return a value from the cache, ignoring the server completely
+  /// (implying that the returned value may be stale with respect to the value on the server). If
+  /// there is no data in the cache to satisfy the [get()] or [getDocuments()] call,
+  /// [DocumentReference.get()] will return an error and [Query.getDocuments()] will return an empty
+  /// [QuerySnapshot] with no documents.
+  cache,
+}
+
+// TODO: this needs to be more precise
+class PlatformQuerySnapshot {
```

Comment:
You should probably model this class on https://firebase.google.com/docs/reference/android/com/google/firebase/firestore/QuerySnapshot

I would remove `asMap()` as well, and make the app-facing plugin use the typed data

Reply:
Yeah, I've been avoiding this type for a while. I have a suspicion that the data that the backend sends back through the channel to the frontend is not *entirely* identical to the one documented, so I'll model this as I go (I might be wrong)